### ISSUE TITLE
where doesn't warn for wrong values

### ISSUE THREAD

**Issue by vipulnsward** [rails/rails#20473]
Currently if we do `User.where(42)` etc. where doesn't warn on the value being passed. 
According to guides we support only conditions, arrays, hashes, strings. 
Everything else just gets converted to `(*somevalue*)` example -

```ruby
irb(main):001:0> User.where 42
  User Load (0.5ms)  SELECT `users`.* FROM `users` WHERE (42)
  => Loads all records
```
This is ok for values apart from integers, (since they raise an error)
```ruby
irb(main):005:0> User.where(true)
RuntimeError: unsupported: TrueClass
```
For integers the query is executed silently. 

Opening here to get more feedback about this behaviour.

**Comment by sjain1107** [comment]
@vipulnsward 

For Rails version 4.1.10
```ruby
2.2.2 :007 > User.where(true)
  User Load (0.5ms)  SELECT `users`.* FROM `users`  WHERE (1)
  => Loads all records
```
For Rails version 4.2.1

```ruby
irb(main):005:0> User.where(true)
RuntimeError: unsupported: TrueClass
```

**Comment by Gaurav2728** [comment]
@vipulnsward Rails 4.1 show all the records of that table, it may cause SQL injection. In 4.2 it show the error thats cool but this kind of error msg should not be show by framework. It should be improve.

**Comment by rafaelfranca** [comment]
Yes, I don't see why accept wrong attributes.

**Comment by rafaelfranca** [comment]
Mind to prepare a PR?

**Comment by vipulnsward** [comment]
Yeah, will take care of it.

**Comment by akihiro17** [comment]
@rafaelfranca @vipulnsward 

Any progress on this issue?

I think the cause for this issue would be that ``` Arel::Visitors::ToSql ``` does not accept TrueClass, while it accepts Fixnum and Bignum.

[Arel::Visitors::ToSql#756](https://github.com/rails/arel/blob/master/lib/arel/visitors/to_sql.rb#L756)
```ruby
alias :visit_TrueClass                     :unsupported

def unsupported o, collector
  raise "unsupported: #{o.class.name}"
end
```

[Arel::Visitors::ToSql#729](https://github.com/rails/arel/blob/master/lib/arel/visitors/to_sql.rb#L729)
```ruby
def literal o, collector; collector << o.to_s; end

alias :visit_Bignum                :literal
alias :visit_Fixnum                :literal
```

To fix this issue, I think an error should be thrown if a numeric value is passed to the ``` where ``` method as shown in the following code.

```ruby
def where(opts = :chain, *rest)
    if opts == :chain
      WhereChain.new(spawn)
    elsif opts.blank?
      self
    elsif opts.is_a?(Numeric)
       raise ArgumentError, 'the value passed to #where must be Array, Hash or String.'
    else
      spawn.where!(opts, *rest)
    end
end
```

What do you think?

**Comment by magikid** [comment]
I agree with @akihiro17. If we throw an error an some types, we should throw them on all types that don't make sense (like Integers).

**Comment by sgrif** [comment]
Fixed by #21944
